### PR TITLE
BLS-12-381 specification

### PR DESCRIPTION
**Pull request by vbuterin**

### CODE REVIEW

File: specs/bls_verify.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+G2_cofactor = 305502333931268344200999753193121504214466019254188142667664032982267604182971884026507427359259977847832272839041616661285803823378372096355777062779109
+field_modulus = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
+
+def hash_to_G2(m):
+    # TODO
+```
+
+
+
+(see
```
Comment (djrtwo) [inline specs/bls_verify.md]: Is this PR ready for review? This last line makes it look like more was still coming

Comment (djrtwo) [reply specs/bls_verify.md]: Oh wait, it is still pending G2 vs. G1 discussion?

Comment (vbuterin) [reply specs/bls_verify.md]: Yeah, it's still pending G2 vs G1.

**Comment by rauljordan** [comment]
For some numbers here related to the G1 vs. G2 discussion:

In our benchmarks for our 12-381 implementation using CGo here: [go-bls](https://github.com/prysmaticlabs/go-bls), we saw:

```bash
2018/11/19 12:04:13 Initialize Pairing took 0.000959 seconds to complete
2018/11/19 12:04:13 HashAndMapToG2 took 0.000298 seconds to complete
```

on average when using Signatures in G1 and Public Keys in G2.

**Comment by axic** [comment]
@schaeff: can you check this, given you implemented something similar with https://github.com/ewasm/ethereum-bls12.rs ?

File: specs/bls_verify.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        return candidate_sqrt / eighth_roots_of_unity[eighth_roots_of_unity.index(check) // 2]
+    return None
+```
+
+`BLSMultiVerify(pubkeys: [uint384], msgs: [bytes32], sig: [uint384], domain: uint64)` is done as follows:
+
+* Verify that each element of `pubkeys` is a valid G1 point and `sig` is a valid G2 point.
+* Convert each element of `mssg` to a G2 point using `hash_to_G2` defined above.
+* Check that the length of `pubkeys` and `msgs` is the same, call the length `L`
+* Do the pairing check: verify `e(pubkeys[0], hash_to_G2(msgs[0])) * ... * e(pubkeys[L-1], hash_to_G2(msgs[L-1])) == e(G1, sig)`
```
Comment (djrtwo) [inline specs/bls_verify.md]: Need to pass `domain` into each `hash_to_G2`

File: specs/bls_verify.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+If `highflag == 3`, the point is the point at infinity and we require `lowflag = x = 0`. Otherwise, we require `highflag == 2`, in which case the point is `(x, y)` where `y` is the valid coordinate such that `(y * 2) // q == lowflag`.
+
+We represent a point in G2 as a pair of 384-bit integers `(z1, z2)` that are each decomposed into `x1`, `highflag1`, `lowflag1`, `x2`, `highflag2`, `lowflag2` as above. We require `lowflag2 == highflag2 == 0`. If `highflag1 == 3`, the point is the point at infinity and we require `lowflag1 == x1 == x2 == 0`. Otherwise, we require `highflag == 2`, in which case the point is `(x1 * i + x2, y)` where `y` is the valid coordinate such that the imaginary part of `y` satisfies `(y_im * 2) // q == lowflag1`.
+
+`BLSVerify(pubkey: uint384, msg: bytes32, sig: [uint384], domain: uint64)` is done as follows:
+
+* Verify that `pubkey` is a valid G1 point and `sig` is a valid G2 point.
+* Convert `msg` to a G2 point using `hash_to_G2` defined below.
+* Do the pairing check: verify `e(pubkey, hash_to_G2(msg)) == e(G1, sig)`
```
Comment (djrtwo) [inline specs/bls_verify.md]: Should clearly define `e` earlier in the spec.

Comment (vbuterin) [reply specs/bls_verify.md]: Done.

File: specs/bls_verify.md (hunk 2 above)
Comment (vbuterin) [reply specs/bls_verify.md]: Done.

**Pull request merged**